### PR TITLE
opentelemetrytracer: Added support to configure a Dynatrace sampler

### LINKED ISSUE #32581: opentelemetrytracer: Add Dynatrace sampler
*Title*: Add Dynatrace custom sampler

*Description*:

Dynatrace has been contributing to Envoy to enhance the OpenTelemetry support. Envoy now has support to export spans via HTTP, can enrich the spans with resource attributes and configure a custom sampler. The only "missing" feature for Dynatrace customers to be able to migrate to a "full OTel solution" in Envoy is a Dynatrace Sampler. 

In a nutshell,  Dynatrace customers can configure the sampling rate in which spans are sampled, thus achieving a dynamic sampling behavior. This allows customers to effectively cost-control by limiting the amount of spans exported without losing important information about their environments.

Thus Dynatrace would like to propose and add a custom sampler to Envoy that is aligned with the our current sampling features, so users using Envoy in a environment where Dynatrace is present, they can have consistent and correct sampling throughout their applications.

### PR DESCRIPTION
**Pull request by joaopgrassi**
Commit Message: Added support to configure a Dynatrace sampler

Additional Description: Given the OpenTracing deprecation, Dynatrace customers need a solution to migrate to the OpenTelememetry tracer in Envoy. We have contributed several features needed and the sampler is the last step. Please refer to the linked issue for more information/context.
Risk Level: Low
Testing: Unit, Integration, Manual
Docs Changes: Proto docs generated and verified locally
Release Notes: Added support to configure a Dynatrace sampler
Platform Specific Features: N/A
[Optional Runtime guard:] N/A
[Optional Fixes #32581] 
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

File: source/extensions/tracers/opentelemetry/samplers/dynatrace/sampler_config_provider.cc [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+
+namespace Envoy {
+namespace Extensions {
+namespace Tracers {
+namespace OpenTelemetry {
+
+SamplerConfigProviderImpl::SamplerConfigProviderImpl(
+    Server::Configuration::TracerFactoryContext& /*context*/,
+    const envoy::extensions::tracers::opentelemetry::samplers::v3::DynatraceSamplerConfig& config)
+    : sampler_config_(config.root_spans_per_minute()) {}
```
Comment (joaopgrassi) [inline source/extensions/tracers/opentelemetry/samplers/dynatrace/sampler_config_provider.cc:14]: Our goal is that sampling configuration is obtained in the background from the Dynatrace API, to achieve dynamic sampling. 

For this PR we went with this simple approach of returning the value from the configuration. The fetching from the API will be added in a follow up PR.

File: api/envoy/extensions/tracers/opentelemetry/samplers/v3/dynatrace_sampler.proto [hunk 2]
```diff
@@ -44,10 +44,10 @@ message DynatraceSamplerConfig {
 
   // Default number of root spans per minute, used when the value can't be obtained from the Dynatrace API.
   //
-  // A default value of `1000` is used when:
+  // A default value of 1000 is used when:
   //
-  // - `root_spans_per_minute` is unset
-  // - `root_spans_per_minute` is set to 0
+  // - root_spans_per_minute is unset
```
Comment (phlax) [inline api/envoy/extensions/tracers/opentelemetry/samplers/v3/dynatrace_sampler.proto:49]:
```suggestion
  // - ``root_spans_per_minute is unset``
```

**Review by phlax — COMMENTED** [review]
these should be literals - so double backticks please

File: api/envoy/extensions/tracers/opentelemetry/samplers/v3/dynatrace_sampler.proto (hunk 2 above)
Comment (joaopgrassi) [reply api/envoy/extensions/tracers/opentelemetry/samplers/v3/dynatrace_sampler.proto]: Ah thanks, I didn't about the double backticks. Had single ones but the build was failing. Fixed in [106992a](https://github.com/envoyproxy/envoy/pull/32598/commits/106992a550961a7d099822e6e95809819836c851)

**Review by mattklein123 — DISMISSED** [review]
Skimmed at a high level and seems fine. Given this is a new opt-in WIP feature I think it's ok to merge and see what issues come up for people that use the feature. Thanks!

**Comment by joaopgrassi** [comment]
@mattklein123 there was a conflict with the changelog.. solved it now :)

**Review by mattklein123 — DISMISSED** [review]

**Comment by joaopgrassi** [comment]
Working to improve the code coverage a bit

**Comment by joaopgrassi** [comment]
@mattklein123 need the approval again, sorry. There was issues with code coverage, but we added more tests and should work now. Thanks!

**Review by mattklein123 — APPROVED** [review]

**Pull request merged**